commerce/bookstore/python/loaders.py

```diff
@@ -0,0 +1,185 @@
+        return typeql_insert_query
+
+
+class RatingInput(Loader):
+    def __init__(self, item):
+        super().__init__(item, config.data_path + "ratings.csv")  # Set exact filename to parse with this class
+
+    def load(self):  # building a TypeQL request to insert a review (review relation)
+        typeql_insert_query = "match $u isa user, has foreign-id '" + self.item["User-ID"] + "'; " \
+                              "$b isa book, has ISBN '" + self.item["ISBN"] + "'; " \
+                              "insert $r (author: $u, product: $b) isa review;" \
+                              "$r has rating " + self.item["Book-Rating"] + ";"
+        return typeql_insert_query
+
+
+class OrderInput(Loader):
+    def __init__(self, item):
+        super().__init__(item, config.data_path + "orders.csv")  # Set exact filename to parse with this class
+
+    def load(self):  # building a TypeQL request to insert an order
+        i = 0
+        typeql_insert_query = "match $u isa user, has foreign-id '" + self.item["User-ID"] + "';"
+        for book in random_books():
+            i += 1  # counter for the number of books
+            if debug: print("Book #" + str(i) + " ISBN: " + book)
+            typeql_insert_query += "$b" + str(i) + " isa book, has ISBN '" + book + "';"
+
+        typeql_insert_query += "insert $o isa order, has id '" + self.item["id"] + "', " \
+                               "has foreign-user-id '" + self.item["User-ID"] + "', has created-date " + self.item["date"] + ", " \
+                               "has status '" + self.item["status"] + "'," \
+                               "has delivery-address '" + self.item["delivery_address"] + "', " \
+                               "has payment-details '" + self.item["payment_details"] + "';"
+        typeql_insert_query += "$o ("
+        for j in range(1, i+1):  # for all i books in the order
+            typeql_insert_query += "item: $b" + str(j) + ","  # Add every book into the relation
+        typeql_insert_query += " author: $u) isa order;"
+
+        return typeql_insert_query
+
+
+class GenreInput(Loader):
+    def __init__(self, item):
+        super().__init__(item, config.data_path + "genres.csv")  # Set exact filename to parse with this class
+
+    def load(self):  # building a TypeQL request to insert a genre/book association
+        typeql_insert_query = "match $b isa book, has ISBN '" + self.item["ISBN"] + "'; " \
+                              "$g isa genre-tag; $g '" + self.item["Genre"] + "'; " \
+                              "insert $b has $g;"
+        return typeql_insert_query
+
+
+def create_genre_tags():  # Creating genre tags and tag hierarchy
+    with TypeDB.core_client("localhost:1729") as client:
+        with client.session(config.db, SessionType.DATA) as session:
+            with session.transaction(TransactionType.WRITE) as transaction:
+                transaction.query().insert("insert $g 'Fiction' isa genre-tag;")
```

Comment:
We should now seek to clean up this code.

Keeping the overall strategy (running a bunch of inserts) is fine; but we need to minimise code duplication. So for example, make a list: `["Fiction", "Non fiction", "Other"]` and so on, and use a for loop to insert all the genre tags - so we don't just duplicate the `transaction.query().insert()` code.

We also need a second list - or perhaps a dictionary: `{"Sci-Fi": "Fiction", "Fantasy": "Fiction", "Biography", "Non fiction"}` etc. This will then allow us to write just a single `match-insert` in our source code, again using a `for` loop.

Finally, given that this is essentially "extra data" that we're creating separately to the CSV files, we should probably define this list and this dictionary in a separate file. Something like: `genre_tags.py`.

Reply:
Fxied in https://github.com/vaticle/typedb-examples/pull/143/commits/ff0e0c69c4d545cd2586d0ed44714b37b603a68e